Support/Fix/Enhancement: Improved Hardpoint Display

RisingLeaf opened the pull request:
**Feature:** This PR implements the feature request detailed and discussed on github

## Feature Details
Basically adds pages to the hardpoint display if there are too many to fit them on one.

![image](https://user-images.githubusercontent.com/85687254/212754723-b1e71254-32f6-4b39-956a-04e40d0bf2c0.png)


## Performance Impact
Slightly increased cpu usage for ships with a lot of hardpoints

RisingLeaf commented on source/ShipInfoPanel.cpp:
```diff
@@ -266,7 +279,10 @@ bool ShipInfoPanel::Drag(double dx, double dy)
 {
 	if(draggingIndex >= 0 && hoverIndex >= 0 && hoverIndex != draggingIndex)
-		(**shipIt).GetArmament().Swap(hoverIndex, draggingIndex);
+	{
```
```suggestion
```

RisingLeaf commented on source/ShipInfoPanel.cpp:
```diff
@@ -266,7 +279,10 @@ bool ShipInfoPanel::Drag(double dx, double dy)
+	{
+			(**shipIt).GetArmament().Swap(hoverRight ? indecesRight[hoverIndex] : indecesLeft[hoverIndex],
+										dragRight ? indecesRight[draggingIndex] : indecesLeft[draggingIndex]);
+	}
```
```suggestion
```

Hurleveur commented on source/ShipInfoPanel.cpp:
```diff
@@ -230,11 +234,20 @@ bool ShipInfoPanel::Click(int x, int y, int /* clicks */)
 	selectedCommodity.clear();
 	selectedPlunder = nullptr;
 	Point point(x, y);
+	if(point.X() <= nextHardpoint.X() + 145. && point.X() >= nextHardpoint.X() + 105.
```
These values should probably be in interface.txt, it's not really important if the rest cannot be customized (havent looked if thats the case), at least we shouldnt add fixed elements

RisingLeaf replied:
Actually after having a deep look, this element is positioned according to how the hardpoints are displayed, which makes it not fixed

Hurleveur replied:
The relative distance is though

RisingLeaf replied:
how can I set this distance in interfaces?
I am not using interfaces.txt very often...

Hurleveur replied:
just look at the mapplanetcard.cpp it uses this extensively

Hurleveur commented on source/ShipInfoPanel.cpp:
```diff
@@ -463,44 +473,152 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
-			topTo = to;
-			topColor = color;
-			hasTop = true;
+			const Hardpoint *hardpoint = weaponsRight[weaponIndex];
```
This code feels like its repeated twice (or almost), it should probably be made into a method, although I'd prefer a lambda function (but those can be tricky to use so do say if you need help)
Such method would take a bool to know if its left or right that we're drawing (which, why do you separate the two btw? it seemed fine to me with the x allowing to know if its left or right)

Hurleveur replied:
Well the lambda can just take the vector as an argument

RisingLeaf replied:
Added a lambda as requested

Hurleveur replied:
Nice! You could always select the elements that are to make it trough and which dont (instead of writing & only writing those you need)
If there are a lot of elements its fine this way for me

Hurleveur reviewed (changes requested):
Havent looked that deep into it because I think we may have to change a few things on how it looks.
This is definitely an improvement, but on the video you placed it still felt a bit too crammed for me, we should probably have the space between weapons defined in interface.txt (and used to know how many pages are needed) also I think that smth to be done to show less is an option somewhere on screen (between the two next and previous buttons? idk I'm no UI expert ^^) to toggle showing only guns/turrets?

vitalchip commented:
I've only skimmed over the code.. is there a way to drag weapons between pages?

Hurleveur commented:
I did not see any, I think that would require temporarily stocking weapons to put them at other pages

vitalchip commented:
Might be possible to drag a weapon over  large arrow at the left or right and while hovering over it the page moves in the appropriate direction.

Having guns and turrets onto seperate pages would help a little since they can't be interchanged anyway but the actual benefit would depend on the gun/turret ratio of the ship and I suspect ships with *lots* of hardpoints would be mostly turrets.

quyykk commented on source/ShipInfoPanel.h:
```diff
@@ -89,14 +90,26 @@ class ShipInfoPanel : public Panel {
+	bool dragRight = false;
+	Point nextHardpoint = Point(0.0, 0.0);
+	Point previousHardpoint = Point(0.0, 0.0);
+	bool nextHover = false;
```
This isn't used anywhere (yet?)

quyykk commented on source/ShipInfoPanel.h:
```diff
@@ -89,14 +90,26 @@ class ShipInfoPanel : public Panel {
+	std::vector<ClickZone<int>> zonesRight;
+	std::vector<ClickZone<int>> zonesLeft;
+	std::vector<const Hardpoint*> weaponsRight;
+	std::vector<int> indecesRight;
```
The plural is indices

quyykk commented on source/ShipInfoPanel.h:
```diff
@@ -89,14 +90,26 @@ class ShipInfoPanel : public Panel {
-	std::vector<ClickZone<int>> zones;
+	std::vector<ClickZone<int>> zonesRight;
+	std::vector<ClickZone<int>> zonesLeft;
+	std::vector<const Hardpoint*> weaponsRight;
```
```suggestion
	std::vector<const Hardpoint *> weaponsRight;
```

quyykk commented on source/ShipInfoPanel.h:
```diff
@@ -89,14 +90,26 @@ class ShipInfoPanel : public Panel {
+	std::vector<ClickZone<int>> zonesLeft;
+	std::vector<const Hardpoint*> weaponsRight;
+	std::vector<int> indecesRight;
+	std::vector<const Hardpoint*> weaponsLeft;
```
```suggestion
	std::vector<const Hardpoint *> weaponsLeft;
```

quyykk commented on source/ShipInfoPanel.h:
```diff
@@ -89,14 +90,26 @@ class ShipInfoPanel : public Panel {
+	std::vector<int> indecesLeft;
+	bool hoverRight = false;
+	bool dragRight = false;
+	Point nextHardpoint = Point(0.0, 0.0);
```
```suggestion
	Point nextHardpoint;
```

quyykk commented on source/ShipInfoPanel.h:
```diff
@@ -89,14 +90,26 @@ class ShipInfoPanel : public Panel {
+	bool hoverRight = false;
+	bool dragRight = false;
+	Point nextHardpoint = Point(0.0, 0.0);
+	Point previousHardpoint = Point(0.0, 0.0);
```
```suggestion
	Point previousHardpoint;
```

quyykk commented on source/ShipInfoPanel.cpp:
```diff
@@ -463,32 +473,93 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
+		index++;
+	}
+
+	auto drawElements = [&] (std::vector<const Hardpoint*> &weaponList, int weaponIndex, bool right) {
```
Lambdas are uppercased, like functions: `DrawElements`. Also a space between `Hardpoint` and `*`. And the { on a new line.

RisingLeaf replied:
okidoki

quyykk commented on source/ShipInfoPanel.cpp:
```diff
@@ -463,32 +473,93 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
 	bool hasTop = false;
 	auto layout = Layout(static_cast<int>(LABEL_WIDTH), Truncate::BACK);
+
+	// First Calculate how many pages are needed. One is always there.
```
```suggestion
	// First calculate how many pages are needed. One is always there.
```

quyykk commented on source/ShipInfoPanel.cpp:
```diff
@@ -463,32 +473,93 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
+	int turretIndex = 0;
+	int rowsPerPage = static_cast<int>((sprite->Height() * scale) / 10.) - 1;
+	int offset = 0;
+	while(pageNeeded)
```
The page calculation can be moved to `UpdateInfo`, because we only need to calculate it once for any given ship.

RisingLeaf replied:
ok

RisingLeaf replied:
took a look, wont work because I need the scale and the sprite to calculate this stuff

quyykk commented:
> Might be possible to drag a weapon over large arrow at the left or right and while hovering over it the page moves in the appropriate direction.

This should probably be done as part of this PR?

RisingLeaf commented:
> > Might be possible to drag a weapon over large arrow at the left or right and while hovering over it the page moves in the appropriate direction.
> 
> This should probably be done as part of this PR?

will do it, when I get to it

RisingLeaf commented:
I still need to do some stuff on the interfaces.txt file

RisingLeaf commented:
Implemented some interface changes

Hurleveur reviewed (commented):
Just tried it manually, moving the hardpoints between different pages works perfectly, the only issue I have is that this uses r and z for no apparent reason (usually we use a key thats within the word, and add _ before it in the word to make it be highlighted when pressing alt)
Also it doesnt seem to work when I click on it

RisingLeaf commented:
I will see why clicking is not working

Hurleveur commented:
r is within p_revious so just rename it to that (adding _ makes it be highlighted)
And then take e and rename next to n_ext?

RisingLeaf commented:
ok, switched keys as proposed and the buttons should now be clickable again.
They werent because the was a zero margin for the y position 🤣

Hurleveur reviewed (approved):
code looks good and since you say it's now works nicely this is good for me
it's almost a bugfix because without it the Augen's many hardpoints dont show up properly

RisingLeaf commented:
Could someone please remove the orphaned tag from this one?

warp-core commented on source/ShipInfoPanel.cpp:
```diff
@@ -449,14 +466,9 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
 	int turretRows = max(count[0][1], count[1][1]);
 	// If there are both guns and turrets, add a gap of ten pixels.
 	double height = 20. * (gunRows + turretRows) + 10. * (gunRows && turretRows);
+	bool pageNeeded = height > (sprite->Height() * scale);
+	height = pageNeeded ? sprite->Height() * scale : height;
```
Why not use the size of the actual bounding box here instead of multiplying the sprite height by scale?
For a particularly wide and short sprite, this will result in this value you're comparing to being much lower than the actual available space; I don't think that's what we want.

warp-core commented on source/ShipInfoPanel.cpp:
```diff
@@ -468,32 +480,95 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
+	pages = 1;
+	int gunIndex = 0;
+	int turretIndex = 0;
+	int rowsPerPage = static_cast<int>((sprite->Height() * scale) / 10.) - 1;
```
Again, `sprite->Height() * scale` may actually be muych less than the actual available height.

warp-core reviewed (commented):
From your screenshots, it looks like you've removed the small gap separating the guns and turrets. And listed turrets in the same row as guns. I don't think you should do this. All the turrets should be below all the guns.

Additionally, from screenshots provided by Hurleveur:
> ![image](https://user-images.githubusercontent.com/94366726/212534866-2e35de14-748c-4eb9-b8bb-b114f3663510.png) second page ![image](https://user-images.githubusercontent.com/94366726/212534876-21816aed-eb1e-4162-95a9-b712f7c2ed1c.png)

Instead of splitting the turrets across two places, with a few at the bottom of a page mostly mad e up of guns and then a few more on their own page, with a lot of empty space left, the turrets should just all be on their own page. This may not always be possible, but where it is, it should be the case.

warp-core commented:
I think another thing to consider is that, if the right column has too many items but the left has free space, some of all central (x = 0) hardpoint labels could be moved to the left.

This might not need to bring so great, because it's be inconsistent, but it might be preferable to using multiple pages where it would allow avoiding that.

Hurleveur commented:
That doesnt seem like a robut system to be, it probably wouldnt support many big plugin ships

Hurleveur commented:
That feels like extra work that could be done in a follow up PR?
It may be more elegant but unnessecary if we have this and it wouldnt support all plugins like this would.

RisingLeaf commented:
Ok, I am not saying the code is clean yet, but the requested things have all been implemented (except the last one, seems to be controversial?)

RisingLeaf commented:
@warp-core I think I addressed most of your feedback. Please have a look.

I do not agree with this:

> I think another thing to consider is that, if the right column has too many items but the left has free space, some of all central (x = 0) hardpoint labels could be moved to the left.
> 
> This might not need to bring so great, because it's be inconsistent, but it might be preferable to using multiple pages where it would allow avoiding that.

What I could do though is just place indices left and right alternating. If they are at x=0. Everything else is a bit out of scope in my opinion.

I also re-added the small gap, but a page break is a bit too much in my opinion, something like that could be done in a follow up.

warp-core commented on source/ShipInfoPanel.cpp:
```diff
@@ -299,13 +321,80 @@ void ShipInfoPanel::UpdateInfo()
+	int turretIndex = 0;
+	rowsPerPage = static_cast<int>(height) / 10. - 1;
+	int offset = 0;
+	while(pageNeeded)
+	{
+		// First go through the guns.
+		if(gunIndex < gunRows)
+		{
+			if((gunRows - gunIndex) * 20. >= height)
+				gunIndex += (rowsPerPage / 2);
+			else
+			{
+				offset = gunRows - gunIndex;
+				gunIndex = gunRows;
+			}
+		}
+		// Then go through the turrets.
+		if(gunIndex >= gunRows && turretIndex < turretRows)
+		{
+			if(offset)
+				if((turretRows - turretIndex + offset) * 20. + 10. >= height)
+					turretIndex += (rowsPerPage / 2) - (offset + 1);
+				else
+					turretIndex = turretRows;
+			else
+			{
+				if((turretRows - turretIndex) * 20. >= height)
+					turretIndex += (rowsPerPage / 2);
+				else
+					pageNeeded = false;
+			}
+		}
+		if(turretIndex >= turretRows && gunIndex >= gunRows)
+			pageNeeded = false;
+		pages++;
+		offset = 0;
+	}
```
This needs some comments.

RisingLeaf replied:
done

RisingLeaf commented on source/ShipInfoPanel.cpp:
```diff
@@ -299,13 +321,80 @@ void ShipInfoPanel::UpdateInfo()
+	// If there are both guns and turrets, add a gap of ten pixels.
+	double height = 20. * (gunRows + turretRows) + 10. * (gunRows && turretRows);
+	bool pageNeeded = height > bounds.Height();
+	height = pageNeeded ? bounds.Height() - 30. : height;
```
If we do pages, we can not use the height of all guns and turrets as a list. We need the height of the bounding box minus space for the buttons

warp-core replied:
Either add a comment saying that or put the amount of height needed by the buttons in a named constant and use that here.

RisingLeaf replied:
added a constant

RisingLeaf commented on source/ShipInfoPanel.cpp:
```diff
@@ -299,13 +321,80 @@ void ShipInfoPanel::UpdateInfo()
+	pages = pageNeeded ? 0 : 1;
+	int gunIndex = 0;
+	int turretIndex = 0;
+	rowsPerPage = static_cast<int>(height) / 10. - 1;
```
In a comment?
Each row takes up 10px minus the gap between guns and turrets

warp-core replied:
What if there is no gap? This assumes that there are both guns and turrets and so there will be a gap, but that might not be the case.
This also seems to count the space on each side of the sprite as separate rows?
Each row takes up 20 pixels, you just have space for two hardpoints (one on each side) in each row.
If this (the number of hardpoints that can be displayed in total on each page) is what this is supposed to be, the variable should be given a more precise name. Something like `hardpointsPerPage`.

RisingLeaf commented on source/ShipInfoPanel.cpp:
```diff
@@ -299,13 +321,80 @@ void ShipInfoPanel::UpdateInfo()
+	int turretRows = max(count[0][1], count[1][1]);
+	// If there are both guns and turrets, add a gap of ten pixels.
+	double height = 20. * (gunRows + turretRows) + 10. * (gunRows && turretRows);
+	bool pageNeeded = height > bounds.Height();
```
renamed

RisingLeaf commented on source/ShipInfoPanel.cpp:
```diff
@@ -15,6 +15,7 @@ this program. If not, see <https://www.gnu.org/licenses/>.
 
 #include "ShipInfoPanel.h"
 
+#include "Hardpoint.h"
```
done

RisingLeaf commented on data/_ui/interfaces.txt:
```diff
@@ -1288,6 +1288,14 @@ interface "info panel"
+	button r "<p_revious"
+		center -95 260.
+		dimensions 60 25
+	
```
done

RisingLeaf commented:
I think I addressed all your stuff again.

(btw please remove the waiting on OP?)

warp-core commented on source/ShipInfoPanel.cpp:
```diff
@@ -299,13 +326,88 @@ void ShipInfoPanel::UpdateInfo()
+		// First go through the guns.
+		if(gunIndex < gunRows)
+		{
+			// The guns still overflow this page.
+			if((gunRows - gunIndex) * 20. >= height)
+				gunIndex += (rowsPerPage / 2);
+			// There is space left after all guns were placed.
+			else
+			{
+				offset = gunRows - gunIndex;
+				gunIndex = gunRows;
+			}
+		}
+		// Then go through the turrets.
+		if(gunIndex >= gunRows && turretIndex < turretRows)
+		{
+			// If guns were already placed on this page.
+			if(offset)
+				// The turrets still exceed this pages space.
+				if((turretRows - turretIndex + offset) * 20. + 10. >= height)
+					turretIndex += (rowsPerPage / 2) - (offset + 1);
+				// The turrets fit this page.
+				else
+					turretIndex = turretRows;
+			else
+			{
+				// The turrets still exceed this pages space.
+				if((turretRows - turretIndex) * 20. >= height)
+					turretIndex += (rowsPerPage / 2);
+				// The turrets fit this page.
+				else
+					overflowsPage = false;
+			}
+		}
+		// Check if we went through all hardpoints.
+		if(turretIndex >= turretRows && gunIndex >= gunRows)
+			overflowsPage = false;
+		pages++;
+		offset = 0;
+	}
```
This doesn't seem like a very good way of going about determining how many pages will be needed.

To begin with, I suggest not using a while loop.
```c++
	if(!overflowsPage)
		return;
```
Because there's nothing else to do in this method if everything fits on one page.

Next, figure out how many full pages of guns we have, and how many remaining guns we'll have afterwards:
```c++
	int gunPages = gunRows / rowsPerPage;
	int remainingGunRows = gunRos % rowsPerPage;
```
Now, let's find out if the turrets will fit on the same page as the remaining guns:
```c++
	bool turretsSharePage = (remainingGunRows + turretRows) < rowsPerPage; // We lose a row for the separation between guns and turrets, so use an exclusive comparison. This will need some slight tweaking in the final product but good enough for this example.
```
If they do fit on the same page, the total number of pages is `gunPages + 1`.
Else, we need to do the same check as we did for guns but for turrets:
```c++
	int turretPages = turretRows / rowsPerPage;
	int remainingTurretRows = turretRows % rowsPerPage;
```
And, then, the total number of pages is going to be:
```c++
	pages = gunPages + turret pages + static_cast<bool>(remainingGunRows) + static_cast<bool>(remainingTurretRows);
```

(I don't know if the code snippets here will actually work. Think of them like pseudocode.)

RisingLeaf replied:
I did something like that, please have a look at it

warp-core commented on source/ShipInfoPanel.cpp:
> The page calculation can be moved to `UpdateInfo`, because we only need to calculate it once for any given ship.

> took a look, wont work because I need the scale and the sprite to calculate this stuff

Why can't you get those things in UpdateInfo?

RisingLeaf replied:
moved

warp-core commented on source/ShipInfoPanel.cpp:
```diff
@@ -468,22 +546,65 @@ void ShipInfoPanel::DrawWeapons(const Rectangle &bounds)
+			gunHardpoints.emplace_back(&hardpoint);
+	}
+
+	// Sort wepaons into left and right.
+	auto SortIntoIndices = [&] (std::vector<const Hardpoint *> &toSort)
+	{
+		for(const Hardpoint *hardpoint : toSort)
+		{
+			if(hardpoint->GetPoint().X() >= 0.)
+			{
+				indicesRight.emplace_back(index);
+				weaponsRight.emplace_back(hardpoint);
+			}
+			else
+			{
+				indicesLeft.emplace_back(index);
+				weaponsLeft.emplace_back(hardpoint);
+			}
+			index++;
+		}
+	};
+	SortIntoIndices(gunHardpoints);
+	SortIntoIndices(turretHardpoints);
```
Wasn't there code already here that sorts the hardpoints into left and right?
Didn't you also sort them into left and right in the other method?
This doesn't seem like a very good way of doing this.
Not only do you seem to be replicating code for doing, if not the same, very similar things across multiple methods, you're running it every frame. This seems especially egregious when the containers you are populating are class fields. It seems counterproductive to store these in the class and then still unnecessarily regenerate them every frame.
The names are also not very informative.

RisingLeaf replied:
Ok, now it is only run alongside UpdateInfo in the same loop as the other stuff, but whats wrong with the names?

RisingLeaf commented:
Just fyi, I wont be available for the next two and half weeks or so, if more changes get requested I will not get to them in this time.

tibetiroka commented:
I think something is broken here
Continuous: 
![image](https://github.com/endless-sky/endless-sky/assets/68112292/9ca07d58-32a8-45d3-8228-8c4990e27b3c)

7453:
![image](https://github.com/endless-sky/endless-sky/assets/68112292/3c6f90c0-e608-4577-9504-89f5a5b1c5b1)

I think all the labels are in the top two position behind each other. I also get weird behaviour when dragging them, where it shows a different name when I drag the label somewhere.

RisingLeaf commented:
Yea this looks broken...
I will look into that.

also all constructive reviews are welcome reviews

RisingLeaf commented:
I did fix an issue with dragging but I was not able to reproduce what I see in your screenshots, more details please and please check if you are on an older commit

tibetiroka commented:
I just did a clean build on the latest version of the branch and still get this issue. The dragging issue was indeed fixed, however.

My save file: [Space Shuttle.txt](https://github.com/endless-sky/endless-sky/files/12919655/Space.Shuttle.txt)

RisingLeaf commented:
thanks for the save file, was able to pin down the bug and I think I fixed it, I would appreciate it if you could check one more time though

tibetiroka commented:
I tested this on b2957590fa40238bd3c301f5cad9e48582d339b1

I only see 4 turrets out of 5, the gridfire in the center seems to be missing still.

![image](https://github.com/endless-sky/endless-sky/assets/68112292/a2cdf1e4-68b9-4587-aa75-6a15c5769ad8)

RisingLeaf commented:
I think that should be it, I hope

warp-core reviewed (changes requested):
This seems to vertically align all the hardpoints to the top instead of the center:

before | after
-- | --
![image](https://github.com/endless-sky/endless-sky/assets/20605679/d4bb0a1d-c36a-4926-b1ff-76eb494e942b) | ![image](https://github.com/endless-sky/endless-sky/assets/20605679/edf89f39-9f39-4d19-b5a9-f8362f658f8c)


It also causes turrets to encroach on the guns if the latter do not appear in equal numbers on both sides:
before | after
-- | --
![image](https://github.com/endless-sky/endless-sky/assets/20605679/b09f7b46-ae7a-47ec-b4c4-a4434e5fb6ae) | ![image](https://github.com/endless-sky/endless-sky/assets/20605679/60fa3d72-b7ec-444b-becf-df4650ef4191)
![image](https://github.com/endless-sky/endless-sky/assets/20605679/8fa0f838-4f30-45c7-b574-12888e90ec56) | ![image](https://github.com/endless-sky/endless-sky/assets/20605679/c50ad861-f8c3-4a9e-bff4-0dd9829435a3)
![image](https://github.com/endless-sky/endless-sky/assets/20605679/9fafdc35-97c0-47fd-9d47-8c518063703a) | ![image](https://github.com/endless-sky/endless-sky/assets/20605679/4c921b14-3b42-46d0-8352-f1d4bd8fd7c0)

RisingLeaf commented on source/ShipInfoPanel.cpp:
```diff
@@ -43,12 +44,19 @@ this program. If not, see <https://www.gnu.org/licenses/>.
+#include <cmath>
+#include <utility>
+
+#include "Logger.h"
```
```suggestion
```

RisingLeaf commented:
My motivation to work on this one is low rn, might get to it in January. Closing for the time being
